pantheon-bundle/src/main/java/com/redhat/pantheon/model/api/util/SafeResourceTraversal.java

```diff
@@ -0,0 +1,91 @@
+package com.redhat.pantheon.model.api.util;
+
+import com.redhat.pantheon.model.api.Child;
+import com.redhat.pantheon.model.api.Field;
+import com.redhat.pantheon.model.api.SlingModel;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static java.util.Optional.ofNullable;
+
+/**
+ * Utility class which offers a null-safe way to traverse deep {@link SlingModel}
+ * structures. Traversals represent paths across multiple nodes ending in deeper
+ * child nodes, or properties. If any of the intermediary nodes is not present, the whole traversal
+ * will still conclude, with a final result of null.<br/>
+ * <br/>
+ * To use it, invoke the {@link SafeResourceTraversal#start(SlingModel)} method with a
+ * model to be traversed. From there on, the whole structure may be traversed all the
+ * way down to specific fields.
+ *
+ * @author Carlos Munoz
+ */
+public class SafeResourceTraversal<T extends SlingModel> implements Supplier<T> {
+
+    private static final SafeResourceTraversal<?> EMPTY = new SafeResourceTraversal<>(null);
+
+    private final Optional<T> currentResource;
+
+    private SafeResourceTraversal(T resource) {
+        this.currentResource = ofNullable(resource);
+    }
+
+    /**
+     * Starts a safe traversal.
+     * @param model The {@link SlingModel} to traverse.
+     * @param <U>
+     * @return A traversal object starting from the given {@link SlingModel}. If the model
+     * is null, traversals will still conclude but will always yield null results.
+     */
+    public static final <U extends SlingModel> SafeResourceTraversal<U> start(@Nullable U model) {
```

Comment:
Can we rename to generic `<U>` to `<M>`? Since the type parameter if model...

Reply:
done.